modyn/supervisor/internal/triggers/utils/decision_policy.py

```diff
@@ -0,0 +1,152 @@
+    `percentile` threshold.
+    """
+
+    def __init__(
+        self,
+        window_size: int,
+        percentile: float,
+        triggering_direction: Literal["higher", "lower"],
+    ):
+        """
+        Args:
+            window_size: The size of the observations to be considered for the decision.
+            percentile: The percentile that a threshold has to be in to trigger event.
+            triggering_direction: Whether a higher score should produce a trigger or a lower score.
+        """
+        super().__init__(window_size, triggering_direction)
+        self.percentile = percentile
+
+    def evaluate_decision(self, measurement: float) -> bool:
+        if len(self.score_observations) == 0:
+            self.score_observations.append(measurement)
+            return True
+
+        sorted_observations = list(sorted(self.score_observations, reverse=self.triggering_direction == "lower"))
+
+        threshold = sorted_observations[
+            min(
+                max(
+                    0,
+                    int(round(len(sorted_observations) * (1.0 - self.percentile))) - 1,  # from length to index space
+                ),
+                len(sorted_observations) - 1,
+            )
+        ]
+        self.score_observations.append(measurement)
+
+        return measurement > threshold
+
+
+class DynamicRollingAverageThresholdPolicy(DynamicDecisionPolicy):
+    """Triggers when a new measurement value deviates from the rolling average
+    by a certain amount or percentage."""
+
+    def __init__(
+        self,
+        window_size: int,
+        deviation: float,
+        absolute: bool,
+        triggering_direction: Literal["higher", "lower"],
+    ):
+        """
+        Args:
+            window_size: The size of the observations to be considered for the decision.
+            deviation: The deviation from the rolling average that triggers.
+            absolute: Whether the deviation is absolute or relative to the rolling average.
+            triggering_direction: Whether a higher score should produce a trigger or a lower score.
+        """
+        super().__init__(window_size, triggering_direction)
+        self.window_size = window_size
+        self.deviation = deviation
+        self.absolute = absolute
+
+    def evaluate_decision(self, measurement: float) -> bool:
+        if not self.score_observations:
+            self.score_observations.append(measurement)
+            return True
+
+        rolling_average = sum(self.score_observations) / len(self.score_observations)
+        deviation = (
+            measurement - rolling_average if self.triggering_direction == "higher" else rolling_average - measurement
+        )
+
+        self.score_observations.append(measurement)
+
+        if self.absolute:
+            return deviation >= self.deviation
+        return deviation >= self.deviation * rolling_average
```

Comment:
I guess we should have a test for the relative case as well (codecov is saying we dont have one)

Reply:
Done